### PR TITLE
feat(oauth): add OAuth endpoint with generateUrl() method

### PR DESCRIPTION
**Pull request by berezovskycom**
This PR is the 2nd part of adding OAuth2.0 into Abstract SDK.
[Link to OAUTH2.0 RFC](https://goabstract.atlassian.net/wiki/spaces/ENG/pages/341705615/OAuth+2.0)
[Link to OAuth interface for SDK](https://goabstract.atlassian.net/wiki/spaces/ENG/pages/447643698/RFC+OAuth+interface+for+JavaScript+SDK#Architecture-Summary)

With this feature, developers are able to generate url for authenticating users to their application.

Example:

```
const abstract = new Abstract.Client({
  clientId,
  clientSecret,
  redirectUri
});

const url = abstract.oauth.generateUrl({
  state: "somestate"
});
```

### CODE REVIEW

File: src/endpoints/Oauth.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// @flow
+import { OAuthCredentialsUndefinedError } from "../errors";
+import type { OAuthURLqueries } from "../types";
+import Endpoint from "../endpoints/Endpoint";
+
+export default class Oauth extends Endpoint {
+  name = "oauth";
+
+  generateUrl(descriptor: OAuthURLqueries) {
```
[the remark this answers is not in the document]
Comment (tommoor) [reply src/endpoints/Oauth.js:9]: `generateAuthorizeUrl`? I think there’s value in the extra prefix as it disambiguates the method somewhat. But agree with consistency

File: src/util/helpers.js [hunk 2]
```diff
@@ -45,3 +51,19 @@ export function wrap(value: any, response?: any) {
...
+
+  if (!clientId || !state || !redirectUri) {
+    throw new BaseError(
+      "Client credentials are missing. Please doublecheck clientId, redirectUri and state"
+    );
+  }
+
+  redirectUri = encodeURIComponent(redirectUri);
+
+  return `https://app.abstract.com/signin/oauth/authorize?client_id=${clientId}&redirect_uri=${redirectUri}&response_type=code&scope=all&state=${state}`;
```
Comment (amccloud) [inline src/util/helpers.js:68]: nit: could inline and avoid re-assignment

```suggestion
  const redirectUri = input.redirectUri || this.options.redirectUri;

  if (!clientId || !state || !redirectUri) {
    throw new BaseError(
      "Client credentials are missing. Please doublecheck clientId, redirectUri and state"
    );
  }

  return `https://app.abstract.com/signin/oauth/authorize?client_id=${clientId}&redirect_uri=${encodeURIComponent(redirectUri)}&response_type=code&scope=all&state=${state}`;
```

Re-assignments and mutations break type refinements. It also has the added benefit of treating the string like a view template 🙂

File: tests/util/helpers.test.js [hunk 3]
```diff
@@ -34,4 +34,18 @@ describe("helpers", () => {
...
   });
+
+  test("generateAuthUrl", () => {
+    const [clientId, redirectUri, state] = ["clientId", "redirectUri", "state"];
+
+    const url = API_CLIENT.generateAuthorizeUrl({
+      clientId,
+      redirectUri,
+      state
+    });
```
Comment (amccloud) [inline tests/util/helpers.test.js:45]: Can also cover not passing one of the values to make sure developers get your intended warning as expected

**Comment by berezovskycom** [comment]
Added all changes to #290. Closing this pr.

**Pull request closed without merging**